### PR TITLE
Visualizations

### PR DESCRIPTION
**Pull request by mstirchak**
Usage: run the django project, and go to localhost:8000/graph.

Generates D3 graphs of db.sead.systems data. Currently, there are two Device buttons, but there is only one working seads device so they both connect to it. Clicking on either Device 1 or Device 2 will allow you to see Daily or Continuous data from the device. Clicking Daily will generate a bar graph with daily power usage/generation, and clicking Continuous will generate a line chart with minute-by-minute power data. The date range for both options can be manually set.

Known bugs: data is taken directly from db.sead.systems, and the server has been on the fritz lately. You may get a network_timed_out error.

Also, the transitions for the Daily bar graphs is bugged. Expanding the date range will not create bars for the new date points. This will be fixed in the next commit. Hopefully.

### CODE REVIEW

File: seadssite/templates/graph.html [hunk 1]
```diff
@@ -0,0 +1,752 @@
+{% extends "base.html" %}
+{% load staticfiles %}
+{% block content %}
+
+<html>
+<meta charset="utf-8">
+  <head>
+    <style>
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Can you put this in a stylesheet?

File: seadssite/templates/graph.html [hunk 2]
```diff
@@ -0,0 +1,752 @@
...
+            .tickFormat(d3.time.format("%Y-%m-%d"));
+
+        var yAxis = d3.svg.axis()
+            .scale(y)
+            .orient("left")
+            .ticks(10);
+
+        var svg = d3.select("body");
+
+/*
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please removed commented out code.

File: seadssite/templates/graph.html [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+       //Update y-axis
+        svg.select(".y.axis")
+           .transition()
+           .duration(1000)
+           .call(yAxis);
+
+
+      var gen_bars = svg.selectAll(".gen_rect")
+            .data(data_gen.power, function(d) { return d.date_gen; });
+            //.data(data_gen.power);
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+          .transition()
+          .duration(500)
+          .attr("x", width)
+          .remove();
+
+          svg.selectAll("bar")
+          .data(data_gen.power)
+          .enter().append("rect")
+          .attr("class", "gen_rect")
+         // .attr("class", "bar")
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+          .style("fill", "steelblue")
+          .attr("x", function(d) { return x(d.date_gen); })
+          .attr("width", x.rangeBand() / 2)
+          .attr("y", function(d) { return y(d.gen/1000000); })
+          .attr("height", function(d) { return height - y(d.gen/1000000); });
+          //.on('mouseover', gen_tip.show)
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 6]
```diff
@@ -0,0 +1,752 @@
...
+        //  .on('mouseout', gen_tip.hide);
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 7]
```diff
@@ -0,0 +1,752 @@
...
+
+          svg.selectAll("bar")
+          .data(data_cons.power)
+          .enter().append("rect")
+          .attr("class", "cons_rect")
+         // .attr("class", "bar")
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 8]
```diff
@@ -0,0 +1,752 @@
...
+          .style("fill", "red")
+          .attr("x", function(d) { return x(d.date_cons) + x.rangeBand()/2; })
+          .attr("width", x.rangeBand()/2)
+          .attr("y", function(d) { return y(d.cons/1000000); })
+          .attr("height", function(d) { return height - y(d.cons/1000000); });
+          //.on('mouseover', use_tip.show)
+          //.on('mouseout', use_tip.hide);
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 9]
```diff
@@ -0,0 +1,752 @@
...
+    function CStart(){
+      var url_cons = Get_Consumed();
+      var url_gen = Get_Generated();
+
+      var margin = {top: 20, right: 20, bottom: 50, left: 50},
+          width = 800 - margin.left - margin.right,
+          height = 400 - margin.top - margin.bottom;
+
+      var parseDate = d3.time.format("%Y-%m-%d %H:%M:%S").parse; // for dates like "2014-01-01"
+      //var parseDate = d3.time.format("%Y-%m-%dT00:00:00Z").parse;  // for dates like "2014-01-01T00:00:00Z"
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 10]
```diff
@@ -0,0 +1,752 @@
...
+        $('#DateRangeCont').hide(500);
+
+        var svg = d3.select("body");
+        svg.selectAll("svg").remove();
+
+        daily = true;
+        continuous = false;
+
+      //  var url_gen = "http://localhost:8000/api/device/466419817/daily/generated_power/1446537600/1446710400?format=json";
+      //  var url_cons = "http://localhost:8000/api/device/466419817/daily/consumed_power/1446537600/1446710400?format=json";
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 11]
```diff
@@ -0,0 +1,752 @@
...
+
+        var url_gen = "http://db.sead.systems:8080/466419817?start_time=1446537600&end_time=1446710400&device=PowerS&granularity=60&list_format=energy";
+        var url_cons = "http://db.sead.systems:8080/466419817?start_time=1446537600&end_time=1446710400&device=Panel1&granularity=60&list_format=energy";
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Why is this hard coded?

File: seadssite/templates/graph.html [hunk 12]
```diff
@@ -0,0 +1,752 @@
...
+            .attr("y", 6)
+            .attr("dy", ".71em")
+            .style("text-anchor", "end")
+            .text("kW");
+
+            svg.selectAll("bar")
+            .data(data_gen.power)
+            .enter().append("rect")
+            .attr("class", "gen_rect")
+           // .attr("class", "bar")
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 13]
```diff
@@ -0,0 +1,752 @@
...
+            .attr("y", function(d) { return y(d.gen/1000000); })
+            .attr("height", function(d) { return height - y(d.gen/1000000); })
+            .on('mouseover', gen_tip.show)
+            .on('mouseout', gen_tip.hide);
+
+            svg.selectAll("bar")
+            .data(data_cons.power)
+            .enter().append("rect")
+            .attr("class", "cons_rect")
+           // .attr("class", "bar")
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 14]
```diff
@@ -0,0 +1,752 @@
...
+        svg.selectAll("svg").remove();
+
+        daily = false;
+        continuous = false;
+      }
+}
+
+    function Continuous(){
+      if(continuous == false || daily == true){
+      //  console.log("Work in Progress!");
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 15]
```diff
@@ -0,0 +1,752 @@
...
+        $('#from').show(500);
+        $('#to').show(500);
+
+        var svg = d3.select("body");
+        svg.selectAll("svg").remove();
+
+        continuous = true;
+        daily = false;
+
+        //var url = "http://localhost:8000/api/device/466419817/daily/generated_power/1446537600/1446710400?format=json";
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 16]
```diff
@@ -0,0 +1,752 @@
...
+        var url = "http://db.sead.systems:8080/466419817?start_time=1446537600&end_time=1446710400&device=Panel1&granularity=60&list_format=energy";
+
+        var margin = {top: 20, right: 20, bottom: 50, left: 50},
+            width = 800 - margin.left - margin.right,
+            height = 400 - margin.top - margin.bottom;
+
+        var parseDate = d3.time.format("%Y-%m-%d %H:%M:%S").parse; // for dates like "2014-01-01"
+        //var parseDate = d3.time.format("%Y-%m-%dT00:00:00Z").parse;  // for dates like "2014-01-01T00:00:00Z"
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 17]
```diff
@@ -0,0 +1,752 @@
...
+
+        var x = d3.time.scale()
+        //.domain([new Date(sample_consumed[0].date), d3.time.day.offset(new Date(sample_consumed[sample_consumed.length - 1].date), 1)])
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 18]
```diff
@@ -0,0 +1,752 @@
...
+            .range([0, width]);
+
+        var y = d3.scale.linear()
+        // .domain([0, d3.max(sample_consumed, function(d) { return +d.power; })])
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 19]
```diff
@@ -0,0 +1,752 @@
...
+        $('#to').show(500);
+
+        var svg = d3.select("body");
+        svg.selectAll("svg").remove();
+
+        continuous = true;
+        daily = false;
+
+        //var url = "http://localhost:8000/api/device/466419817/daily/generated_power/1446537600/1446710400?format=json";
+        var url = "http://db.sead.systems:8080/466419817?start_time=1446537600&end_time=1446710400&device=Panel1&granularity=60&list_format=energy";
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Why is this hard coded?

File: seadssite/templates/graph.html [hunk 20]
```diff
@@ -0,0 +1,752 @@
...
+
+          svg.append("path")
+              .datum(data.power)
+              .attr("class", "line")
+              .attr("d", line);
+          });
+
+    }
+      else{
+      //  console.log("Work in Progress!");
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 21]
```diff
@@ -0,0 +1,752 @@
...
+        continuous = false;
+        daily = false;
+      }
+    }
+
+   function Get_Generated(){
+    var start = getStart();
+    var end = getEnd();
+    //console.log(start);
+    //console.log(end);
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 22]
```diff
@@ -0,0 +1,752 @@
...
+
+    var url = "http://db.sead.systems:8080/466419817?start_time=" + start + "&end_time=" + end + "&device=PowerS&granularity=3600&list_format=energy";
+    //console.log(url);
+      //  var svg = d3.select("body");
+      //  svg.selectAll("svg").remove();
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 23]
```diff
@@ -0,0 +1,752 @@
...
+   function Get_Minute(){
+    var start = getStart();
+    var end = getEnd();
+    //console.log(start);
+    //console.log(end);
+
+    var url = "http://db.sead.systems:8080/466419817?start_time=" + start + "&end_time=" + end + "&device=Panel1&granularity=60&list_format=energy";
+    //console.log(url);
+      //  var svg = d3.select("body");
+      //  svg.selectAll("svg").remove();
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 24]
```diff
@@ -0,0 +1,752 @@
...
+        //svg.selectAll("svg").remove();
+
+    return url;
+   }
+
+   function Get_Minute(){
+    var start = getStart();
+    var end = getEnd();
+    //console.log(start);
+    //console.log(end);
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 25]
```diff
@@ -0,0 +1,752 @@
...
+   function Get_Consumed(){
+    var start = getStart();
+    var end = getEnd();
+    //console.log(start);
+    //console.log(end);
+
+    var url = "http://db.sead.systems:8080/466419817?start_time=" + start + "&end_time=" + end + "&device=Panel1&granularity=3600&list_format=energy";
+    //console.log(url);
+        //var svg = d3.select("body");
+        //svg.selectAll("svg").remove();
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 26]
```diff
@@ -0,0 +1,752 @@
...
+      //  svg.selectAll("svg").remove();
+
+    return url;
+   }
+
+   function Get_Consumed(){
+    var start = getStart();
+    var end = getEnd();
+    //console.log(start);
+    //console.log(end);
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 27]
```diff
@@ -0,0 +1,752 @@
...
+    $('#daily').hide();
+    $('#continuous').hide();
+
+    var daily = false;
+    var continuous = false;
+
+    function Start(){
+        var url_min = Get_Minute();
+      //var url = Refresh();
+      //console.log(url);
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 28]
```diff
@@ -0,0 +1,752 @@
...
+    <script src="http://d3js.org/d3.v3.js"></script>
+    <script src="http://labratrevenge.com/d3-tip/javascripts/d3.tip.v0.6.3.js"></script>
+    <script>
+
+    $('label[for="from"]').hide();
+    $('label[for="to"]').hide();
+    $('#from').hide();
+    $('#to').hide();
+    $('#DateRange').hide();
+    $('#DateRangeCont').hide();
```
Comment (iangudger) [inline seadssite/templates/graph.html]: This code seems to be repeated a few times. Maybe make it a function?

File: seadssite/templates/graph.html [hunk 29]
```diff
@@ -0,0 +1,752 @@
...
+           .attr("x", function(d) { return x(d.date_gen); })
+           .attr("y", function(d) { return y(d.gen/1000000) })
+           .attr("width", x.rangeBand()) //returns rangeRoundBands width
+           .attr("height", function(d) { return height - y(d.gen/1000000) });
+
+
+       //Update rectangles
+       gen_bars.transition()
+           .duration(1000)
+           //.style("fill", function(d,i) { return color(i) })
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 30]
```diff
@@ -0,0 +1,752 @@
...
+          .attr("x", function(d) { return x(d.date_cons) + x.rangeBand()/2; })
+          .attr("y", function(d) { return y(d.cons/1000000) })
+          .attr("width", x.rangeBand() / 2) //returns rangeRoundBands width
+          .attr("height", function(d) { return height - y(d.cons/1000000) });
+
+
+      //Update rectangles
+      cons_bars.transition()
+          .duration(1000)
+          //.style("fill", function(d,i) { return color(i) })
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 31]
```diff
@@ -0,0 +1,752 @@
...
+           .remove();
+
+      var cons_bars = svg.selectAll(".cons_rect")
+            .data(data_cons.power, function(d) { return d.date_cons; });
+            //.data(data_cons.power);
+
+      //Enter rectangles
+      cons_bars.enter()
+          .append("rect")
+          //.style("fill", function(d,i) { return color(i) })
```
Comment (iangudger) [inline seadssite/templates/graph.html]: Please remove commented out code.

File: seadssite/templates/graph.html [hunk 32]
```diff
@@ -0,0 +1,752 @@
...
+
+
+      //Update rectangles
+      cons_bars.transition()
+          .duration(1000)
+          //.style("fill", function(d,i) { return color(i) })
+          .attr("x", function(d) { return x(d.date_cons) + x.rangeBand()/2; })
+          .attr("y", function(d) { return y(d.cons/1000000) })
+          .attr("width", x.rangeBand() / 2) //returns rangeRoundBands width
+          .attr("height", function(d) { return height - y(d.cons/1000000) });
```
Comment (iangudger) [inline seadssite/templates/graph.html]: This code seems to be repeated a few times. Please try to make it a function.

**Comment by iangudger** [comment]
There is a lot of repeated code. Please try to dedupe it with functions.

File: seadssite/templates/graph.html (hunk 1 above)
Comment (ianlofs) [inline seadssite/templates/graph.html]: @mstirchak do you plan on adding this?

File: seadssite/templates/graph.html [hunk 33]
```diff
@@ -0,0 +1,737 @@
...
+
+    }
+
+    function Update_Daily_Graph(panel){
+        if(panel == 1)
+           var url_cons = Get_Panel1_Daily();
+        else if(panel == 2)
+           var url_cons = Get_Panel2_Daily();
+        else if(panel == 3)
+           var url_cons = Get_Panel3_Daily();
```
Comment (ianlofs) [inline seadssite/templates/graph.html]: these Get_Panel functions can all be combined into one function like this, 

``` js
function getPanelUrl(start, end, device, granularity) {
    return "http://db.sead.systems:8080/466419817?start_time=" + start + "&end_time=" + end + "&list_format=energy&type=P&device=" + device + "&granularity=" + granularity;
} 
```

File: seadssite/templates/graph.html [hunk 34]
```diff
@@ -0,0 +1,737 @@
...
+                  });
+                  data_p1.data.forEach(function(d) {
+                    d.date_p1 = parseDate(d.time);
+                    d.p1 = +d.energy;
+                  });
+
+        x.domain(data_gen.data.map(function(d) { return d.date_gen }));
+        y.domain([0, d3.max(data_gen.data, function(d) { return d.gen; })]);
+
+          svg.append("g")
```
Comment (ianlofs) [inline seadssite/templates/graph.html]: indent

**Comment by ianlofs** [comment]
@iangudger did some basic deduping and added fixmes where necessary, going to merge this because we have a demo tomorrow.

**Pull request merged**